src/Adapter/Pgsql.php

```diff
@@ -12,4 +15,47 @@ public function setClientTimezone(Connection $db)
 
         return $this;
     }
+
+    public function registerQueryBuilderCallbacks(QueryBuilder $queryBuilder)
+    {
+        $queryBuilder->on(QueryBuilder::ON_ASSEMBLE_SELECT, function (Select $select) {
+            $groupBy = $select->getGroupBy();
+            if (! empty($groupBy)) {
+                // All SELECT columns must appear in the GROUP BY clause or be used in an aggregate function.
+                $candidates = [];
+                foreach ($select->getColumns() as $alias => $column) {
+                    if ($column instanceof Expression) {
+                        // Assume an aggregate function here.
+                        continue;
+                    }
+
+                    $candidates[] = is_int($alias) ? $column : $alias;
+                }
+
+                $select->groupBy(array_diff($candidates, $groupBy));
+            }
+
+            if ($select->getDistinct() && $select->hasOrderBy()) {
+                // For SELECT DISTINCT, ORDER BY expressions must appear in SELECT list.
+                $candidates = [];
+                foreach ($select->getOrderBy() as list($column, $_)) {
+                    if ($column instanceof Expression) {
+                        // TODO(lippserd): I'm not sure about PostgreSQL's interpretation here, but since expressions
+                        // can be anything, including non-aggregate functions, such expressions must be in the SELECT
+                        // list and referenced manually with an alias in ORDER BY.
```

Comment:
Would this then here not be a string? So, if it's still an expression, throw an error?

Reply:
I don't know how PostgreSQL behaves here or what it would complain about using expressions (I don't think any test would cover 100% of the possibilities), but since they are allowed to order in our library, throwing an exception is too much I think.